code/modules/goonchat/browserOutput.dm

```diff
@@ -0,0 +1,289 @@
+	//Check for existing chat
+	if(!owner)
+		return FALSE
+
+	if(!winexists(owner, "browseroutput")) // Oh goddamnit.
+		alert(owner.mob, "Updated chat window does not exist. If you are using a custom skin file please allow the game to update.")
+		broken = TRUE
+		return FALSE
+
+	if(winget(owner, "browseroutput", "is-disabled") == "false") //Already setup
+		doneLoading()
+
+	else //Not setup
+		load()
+
+	return TRUE
+
+/datum/chatOutput/proc/load()
+	set waitfor = FALSE
+	if(!owner)
+		return
+
+	var/static/list/chatResources = list(
+		"code/modules/html_interface/js/jquery.min.js",
+		"code/modules/goonchat/browserassets/js/json2.min.js",
+		"code/modules/goonchat/browserassets/js/browserOutput.js",
+		"tgui/assets/fonts/fontawesome-webfont.eot",
+		"tgui/assets/fonts/fontawesome-webfont.svg",
+		"tgui/assets/fonts/fontawesome-webfont.ttf",
+		"tgui/assets/fonts/fontawesome-webfont.woff",
+		"code/modules/goonchat/browserassets/css/font-awesome.css",
+		"code/modules/goonchat/browserassets/css/browserOutput.css"
+	)
+
+	// to_chat(world.log, "chatOutput: load()")
+	for(var/attempts in 1 to 5)
+		for(var/asset in chatResources)
+			owner << browse_rsc(file(asset))
+
+		//log_world("Sending main chat window to client [owner.ckey]")
+		owner << browse(file("code/modules/goonchat/browserassets/html/browserOutput.html"), "window=browseroutput")
+		sleep(14 + (chatResources.len * 7))
+		if(!owner || loaded)
+			break
+
+	if(owner && !loaded)
+		doneLoading() // try doing this manually
+		CRASH("[owner] failed to load chat. Attempting doneLoading() manually")
+	// log_world("chatOutput: [owner.ckey] load() completed")
+
+/datum/chatOutput/Topic(href, list/href_list)
+	if(usr.client != owner)
+		return TRUE
+
+	// Build arguments.
+	// Arguments are in the form "param[paramname]=thing"
+	var/list/params = list()
+	for(var/key in href_list)
+		if(length(key) > 7 && findtext(key, "param")) // 7 is the amount of characters in the basic param key template.
+			var/param_name = copytext(key, 7, -1)
+			var/item       = href_list[key]
+
+			params[param_name] = item
+
+	var/data // Data to be sent back to the chat.
+	switch(href_list["proc"])
+		if("doneLoading")
+			data = doneLoading(arglist(params))
+
+		if("debug")
+			data = debug(arglist(params))
+
+		if("ping")
+			data = ping(arglist(params))
+
+		if("analyzeClientData")
+			data = analyzeClientData(arglist(params))
+
+	if(data)
+		ehjax_send(data = data)
+
+//Called on chat output done-loading by JS.
+/datum/chatOutput/proc/doneLoading()
+	if(loaded)
+		return
+
+	loaded = TRUE
+	winset(owner, "browseroutput", "is-disabled=false")
+	for(var/message in messageQueue)
+		to_chat(owner, message)
+
+	messageQueue = null
+	sendClientData()
+
+	pingLoop()
+
+/datum/chatOutput/proc/pingLoop()
+	set waitfor = FALSE
+
+	while (owner)
+		ehjax_send(data = owner.is_afk(29) ? "softPang" : "pang") // SoftPang isn't handled anywhere but it'll always reset the opts.lastPang.
+		sleep(30)
```

Comment:
This should be called by the server_maint subsystem, but, like I said, it's not needed